### PR TITLE
Several bug fixes and improvements

### PR DESCRIPTION
**Pull request by edfelten**
This PR rolls up several bug fixes and enhancements.
* Implement the `extcodehash` EVM opcode
* Generate code segments for Ethereum precompiles correctly
* Make error debugprints more informative
* Handle cross-contract calls to non-existent addresses correctly
* Truncate to 20 bytes all addresses received from application contract code
* In Rust app, add command line option to generate step-by-step execution trace

### CODE REVIEW

File: arb_os/evmCallStack.mini [hunk 1]
```diff
@@ -322,81 +320,91 @@ public impure func evmCallStack_doCall(
...
+            errorHandler();  // should never return
+            panic;  // callcode call type not yet implemented
+        }
 
-    // there were no errors in the setup, so write the result to global storage
-    globalCallStack = Some(newTopFrame);
+        let callerAddr = topFrame.runningAs;
+        if (callerAddr == calleeAddr) {
+            return None;  // a contract shouldn't use the external call mechanism to call itself
+        }
```
Comment (hkalodner) [inline arb_os/evmCallStack.mini:341]: A contract should actually be able to do this

File: arb_os/evmCallStack.mini [hunk 2]
```diff
@@ -322,81 +320,91 @@ public impure func evmCallStack_doCall(
...
+
+        // now create a new callframe for the callee
+        //     it incorporates the balance transfer from caller to callee
+        //          (which is put into the callee frame so it isn't committed unless the call succeeds)
+        let runningAsAddr = calleeAddr;
+        if (kind == 2) {
+            runningAsAddr = callerAddr;  // delegatecall runs as the caller, with callee's code
+        }
+        let updatedAcctStore = topFrame.accountStore;
+        if (kind == 0) {  //TODO: verify that txcall is the only one that is supposed to allow payment-with-call
```
Comment (hkalodner) [inline arb_os/evmCallStack.mini:367]: Can you clarify what this TODO means

File: src/run/emulator.rs [hunk 3]
```diff
@@ -504,13 +517,14 @@ impl Machine {
...
             total_gas_usage: Uint256::zero(),
+            trace_writer: None,
         }
     }
 
     ///Pushes 0 to the stack and sets the program counter to the first instruction. Used by the EVM
     /// compiler.
     pub fn start_at_zero(&mut self) {
-        self.stack.push_usize(0);
+        //self.stack.push_usize(0);
```
Comment (hkalodner) [inline src/run/emulator.rs:527]: This can probably juts be deleted

Comment (edfelten) [reply src/run/emulator.rs:527]: Done.

File: arb_os/evmCallStack.mini (hunk 2 above)
Comment (edfelten) [reply arb_os/evmCallStack.mini:367]: This is now resolved.  Code now handles all of these cases correctly.

File: arb_os/evmCallStack.mini (hunk 1 above)
Comment (edfelten) [reply arb_os/evmCallStack.mini:341]: Removed this code.

**Pull request merged**